rclpy/src/rclpy/context.cpp

```diff
@@ -30,93 +30,93 @@
 
 namespace rclpy
 {
-size_t
-context_get_domain_id(py::capsule pycontext)
+Context::Context()
 {
-  auto context = static_cast<rcl_context_t *>(
-    rclpy_handle_get_pointer_from_capsule(pycontext.ptr(), "rcl_context_t"));
-  if (!context) {
-    throw py::error_already_set();
-  }
+  // Create a client
+  rcl_context_ = std::shared_ptr<rcl_context_t>(
+    new rcl_context_t,
+    [](rcl_context_t * context)
+    {
+      if (NULL != context->impl) {
+        rcl_ret_t ret;
+        if (rcl_context_is_valid(context)) {
+          // shutdown first, if still valid
+          ret = rcl_shutdown(context);
+          if (RCL_RET_OK != ret) {
+            RCUTILS_SAFE_FWRITE_TO_STDERR(
+              "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+              "failed to shutdown rcl_context_t");
+            RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
+            RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
```

Comment:
I recommend using Python warnings instead of writing to stderr

https://github.com/ros2/rclpy/blob/2bb0a8065d437668b145a71c165193365b116312/rclpy/src/rclpy/client.cpp#L68-L74